File: sdk/android/src/main/kotlin/com/itoken/team/nativerpc/core/NativeRPCHost.kt
// NativeRPCHost.kt
// NativeRPC v2
//
// Central RPC host that manages services, routing, and connections
//
// Protocol (simplified JSON-RPC 2.0):
// Request:      {"id": "1", "method": "service.method", "params": {...}}
// Response:     {"id": "1", "result": ...}
// Error:        {"id": "1", "error": {"code": -32601, "message": "..."}}
// Notification: {"method": "service.event", "params": {...}}

package com.itoken.team.nativerpc.core

import com.itoken.team.nativerpc.connection.NativeRPCConnection
import kotlinx.coroutines.*
import kotlinx.serialization.json.*
import java.util.concurrent.ConcurrentHashMap

/**
 * The central RPC host that manages services, handles message routing,
 * and maintains connections to clients.
 *
 * Usage:
 * ```kotlin
 * val host = NativeRPCHost()
 *
 * // Register services
 * host.register(AppService())
 * host.register(UserService())
 *
 * // Connect to Flutter
 * val connection = FlutterMethodChannelConnection(channel)
 * host.addConnection(connection)
 * ```
 */
class NativeRPCHost : NativeRPCHostProtocol {
    
    // MARK: - Properties
    
    /** Registered services by name */
    private val services = ConcurrentHashMap<String, NativeRPCService>()
    
    /** Active connections */
    private val connections = mutableListOf<NativeRPCConnection>()
    private val connectionsLock = Any()
    
    /** Event subscriptions: [fullEventName: Set<connectionId>] */
    private val subscriptions = ConcurrentHashMap<String, MutableSet<String>>()
    
    /** Coroutine scope for async operations */
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    /** JSON serializer */
    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }
    
    // MARK: - Service Registration
    
    /**
     * Register a service with the host.
     *
     * @param service The service to register
     */
    fun register(service: NativeRPCService) {
        val name = service.name
        
        if (services.containsKey(name)) {
            println("[NativeRPC] Warning: Service '$name' is already registered, replacing...")
        }
        
        services[name] = service
        service.onRegistered(this)
        
        println("[NativeRPC] Registered service: $name")
    }
    
    /**
     * Register multiple services at once
     */
    fun register(vararg services: NativeRPCService) {
        for (service in services) {
            register(service)
        }
    }
    
    /**
     * Unregister a service by name
     */
    fun unregister(name: String) {
        services.remove(name)?.let { service ->
            service.destroy()
            // Remove all subscriptions for this service
            subscriptions.keys.filter { it.startsWith("$name.") }.forEach { subscriptions.remove(it) }
            println("[NativeRPC] Unregistered service: $name")
        }
    }
    
    /**
     * Get a registered service by name
     */
    override fun getService(name: String): NativeRPCService? {
        return services[name]
    }
    
    // MARK: - Connection Management
    
    /**
     * Add a connection to the host
     */
    fun addConnection(connection: NativeRPCConnection) {
        synchronized(connectionsLock) {
            connections.add(connection)
            connection.onMessage = { data ->
                handleIncomingMessage(data, connection)
            }
            println("[NativeRPC] Added connection: ${connection.id}")
        }
    }
    
    /**
     * Remove a connection from the host
     */
    fun removeConnection(connection: NativeRPCConnection) {
        synchronized(connectionsLock) {
            connections.removeAll { it.id == connection.id }
            
            // Remove all subscriptions for this connection
            for ((_, connectionIds) in subscriptions) {
                connectionIds.remove(connection.id)
            }
            
            println("[NativeRPC] Removed connection: ${connection.id}")
        }
    }
    
    // MARK: - Message Handling
    
    /**
     * Handle an incoming message from a connection
     */
    private fun handleIncomingMessage(data: String, connection: NativeRPCConnection) {
        scope.launch {
            try {
                val incomingMessage = parseMessage(data)
                
                when (incomingMessage) {
                    is NativeRPCIncomingMessage.Call -> {
                        handleCallRequest(incomingMessage.request, connection)
                    }
                    is NativeRPCIncomingMessage.Subscribe -> {
                        handleSubscribe(incomingMessage.request, connection)
                    }
                    is NativeRPCIncomingMessage.Unsubscribe -> {
                        handleUnsubscribe(incomingMessage.request, connection)
                    }
                }
            } catch (e: NativeRPCError) {
                sendError("unknown", e, connection)
            } catch (e: Exception) {
                val rpcError = NativeRPCError.parseError(e.message)
                sendError("unknown", rpcError, connection)
            }
        }
    }
    
    /**
     * Parse incoming JSON message and determine its type
     */
    private fun parseMessage(data: String): NativeRPCIncomingMessage {
        val jsonObject = try {
            json.parseToJsonElement(data).jsonObject
        } catch (e: Exception) {
            throw NativeRPCError.parseError("Invalid JSON")
        }
        
        // Check for id (required for requests)
        val id = jsonObject["id"]?.jsonPrimitive?.contentOrNull
            ?: throw NativeRPCError.invalidRequest("Missing id in request")
        
        // Check for method (required)
        val method = jsonObject["method"]?.jsonPrimitive?.contentOrNull
            ?: throw NativeRPCError.invalidRequest("Missing method in request")
        
        val params = jsonObject["params"]
        
        // Check for special RPC methods
        when (method) {
            "rpc.subscribe" -> {
                val event = (params as? JsonObject)?.get("event")?.jsonPrimitive?.contentOrNull
                    ?: throw NativeRPCError.invalidParams("rpc.subscribe requires params.event")
                return NativeRPCIncomingMessage.Subscribe(NativeRPCSubscribeRequest(id, event))
            }
            "rpc.unsubscribe" -> {
                val event = (params as? JsonObject)?.get("event")?.jsonPrimitive?.contentOrNull
                    ?: throw NativeRPCError.invalidParams("rpc.unsubscribe requires params.event")
                return NativeRPCIncomingMessage.Unsubscribe(NativeRPCUnsubscribeRequest(id, event))
            }
            else -> {
                // Regular method call
                return NativeRPCIncomingMessage.Call(NativeRPCRequest(id, method, params))
            }
        }
    }
    
    /**
     * Handle a call request
     */
    private suspend fun handleCallRequest(request: NativeRPCRequest, connection: NativeRPCConnection) {
        val serviceName = request.service
        val methodName = request.methodName
        
        // Find the service
        val service = services[serviceName]
        if (service == null) {
            val error = NativeRPCError.serviceNotFound(serviceName)
            sendError(request.id, error, connection)
            return
        }
        
        // Check if method exists
        if (!service.canHandle(methodName)) {
            val error = NativeRPCError.methodNotFound(methodName, serviceName)
            sendError(request.id, error, connection)
            return
        }
        
        // Convert params to args list
        val args = paramsToArgs(request.params)
        
        try {
            // Call the method
            val result = service.handleCall(methodName, args)
            
            // Send success response
            val response = NativeRPCResponse.success(request.id, anyToJsonElement(result))
            sendResponse(response, connection)
        } catch (e: NativeRPCError) {
            sendError(request.id, e, connection)
        } catch (e: Exception) {
            val rpcError = NativeRPCError.internalError(e.message)
            sendError(request.id, rpcError, connection)
        }
    }
    
    /**
     * Convert params (JsonElement) to args list
     */
    private fun paramsToArgs(params: JsonElement?): List<Any?> {
        return when (params) {
            null -> emptyList()
            is JsonArray -> params.map { jsonElementToAny(it) }
            is JsonObject -> listOf(jsonElementToAny(params))
            else -> listOf(jsonElementToAny(params))
        }
    }
    
    /**
     * Handle a subscribe request
     */
    private fun handleSubscribe(request: NativeRPCSubscribeRequest, connection: NativeRPCConnection) {
        val serviceName = request.service
        val eventName = request.eventName
        val fullEventName = request.event
        
        // Validate service exists
        val service = services[serviceName]
        if (service == null) {
            val error = NativeRPCError.serviceNotFound(serviceName)
            sendError(request.id, error, connection)
            return
        }
        
        // Validate event is declared
        if (!service.definitionContainer.hasEvent(eventName)) {
            val error = NativeRPCError.eventNotDeclared(eventName, serviceName)
            sendError(request.id, error, connection)
            return
        }
        
        // Add subscription
        val eventSubscribers = subscriptions.getOrPut(fullEventName) { mutableSetOf() }
        
        val isFirstSubscriber = eventSubscribers.isEmpty()
        eventSubscribers.add(connection.id)
        
        // Notify service if this is the first subscriber
        if (isFirstSubscriber) {
            service.onStartObserving(eventName)
        }
        
        // Send success response
        val response = NativeRPCResponse.success(request.id, JsonPrimitive(true))
        sendResponse(response, connection)
    }
    
    /**
     * Handle an unsubscribe request
     */
    private fun handleUnsubscribe(request: NativeRPCUnsubscribeRequest, connection: NativeRPCConnection) {
        val serviceName = request.service
        val eventName = request.eventName
        val fullEventName = request.event
        
        val eventSubscribers = subscriptions[fullEventName]
        eventSubscribers?.remove(connection.id)
        
        val noMoreSubscribers = eventSubscribers?.isEmpty() == true
        
        // Notify service if no more subscribers
        if (noMoreSubscribers) {
            services[serviceName]?.onStopObserving(eventName)
        }
        
        // Send success response
        val response = NativeRPCResponse.success(request.id, JsonPrimitive(true))
        sendResponse(response, connection)
    }
    
    // MARK: - Event Sending (NativeRPCHostProtocol)
    
    /**
     * Send an event (notification) to all subscribers
     */
    override fun sendEvent(notification: NativeRPCNotification) {
        // Find subscribers for this event
        val subscriberIds = subscriptions[notification.method] ?: return
        
        // Encode the notification
        val data = try {
            json.encodeToString(NativeRPCNotification.serializer(), notification)
        } catch (e: Exception) {
            println("[NativeRPC] Failed to encode notification: ${notification.method}")
            return
        }
        
        // Send to all subscribed connections
        synchronized(connectionsLock) {
            for (connection in connections) {
                if (subscriberIds.contains(connection.id)) {
                    connection.send(data)
                }
            }
        }
    }
    
    /**
     * Send an event with service, event name and data
     */
    fun sendEvent(service: String, event: String, params: JsonElement? = null) {
        val notification = NativeRPCNotification.create(service, event, params)
        sendEvent(notification)
    }
    
    // MARK: - Response Helpers
    
    private fun sendResponse(response: NativeRPCResponse, connection: NativeRPCConnection) {
        val data = try {
            json.encodeToString(NativeRPCResponse.serializer(), response)
        } catch (e: Exception) {
            println("[NativeRPC] Failed to encode response")
            return
        }
        connection.send(data)
    }
    
    private fun sendError(id: String, error: NativeRPCError, connection: NativeRPCConnection) {
        val response = NativeRPCErrorResponse.from(id, error)
        val data = try {
            json.encodeToString(NativeRPCErrorResponse.serializer(), response)
        } catch (e: Exception) {
            println("[NativeRPC] Failed to encode error response")
            return
        }
        connection.send(data)
    }
    
    // MARK: - Lifecycle
    
    /**
     * Notify all services that app entered foreground
     */
    fun onAppForeground() {
        for (service in services.values) {
            service.onForeground()
        }
    }
    
    /**
     * Notify all services that app entered background
     */
    fun onAppBackground() {
        for (service in services.values) {
            service.onBackground()
        }
    }
    
    /**
     * Shutdown the host and all services
     */
    fun shutdown() {
        // Cancel coroutine scope
        scope.cancel()
        
        // Destroy all services
        for (service in services.values) {
            service.destroy()
        }
        services.clear()
        
        // Close all connections
        synchronized(connectionsLock) {
            for (connection in connections) {
                connection.close()
            }
            connections.clear()
        }
        
        // Clear subscriptions
        subscriptions.clear()
        
        println("[NativeRPC] Host shutdown complete")
    }
    
    // MARK: - Introspection
    
    /**
     * Get list of all registered service names
     */
    fun getServiceNames(): List<String> {
        return services.keys.toList()
    }
    
    /**
     * Get service info for introspection
     */
    fun getServiceInfo(name: String): ServiceInfo? {
        val service = services[name] ?: return null
        return ServiceInfo(
            name = name,
            methods = service.definitionContainer.getMethodNames(),
            events = service.definitionContainer.getEventNames().toList(),
            constants = service.getConstants().keys.toList()
        )
    }
    
    // MARK: - JSON Helpers
    
    private fun jsonElementToAny(element: JsonElement): Any? {
        return when (element) {
            is JsonNull -> null
            is JsonPrimitive -> {
                when {
                    element.isString -> element.content
                    element.content == "true" -> true
                    element.content == "false" -> false
                    element.content.contains('.') -> element.content.toDoubleOrNull()
                    else -> element.content.toLongOrNull() ?: element.content.toIntOrNull() ?: element.content
                }
            }
            is JsonArray -> element.map { jsonElementToAny(it) }
            is JsonObject -> element.mapValues { jsonElementToAny(it.value) }
        }
    }
    
    private fun anyToJsonElement(value: Any?): JsonElement? {
        return when (value) {
            null -> null
            is String -> JsonPrimitive(value)
            is Number -> JsonPrimitive(value)
            is Boolean -> JsonPrimitive(value)
            is Map<*, *> -> {
                val map = value.entries.associate { 
                    (it.key as String) to (anyToJsonElement(it.value) ?: JsonNull)
                }
                JsonObject(map)
            }
            is List<*> -> {
                val list = value.map { anyToJsonElement(it) ?: JsonNull }
                JsonArray(list)
            }
            else -> JsonPrimitive(value.toString())
        }
    }
}

// MARK: - Supporting Types

/**
 * Service information for introspection
 */
data class ServiceInfo(
    val name: String,
    val methods: List<String>,
    val events: List<String>,
    val constants: List<String>
)

File: sdk/ios/Sources/NativeRPCKit/Core/NativeRPCHost.swift
// NativeRPCHost.swift
// NativeRPC v2
//
// Central RPC host that manages services, routing, and connections
// Protocol: Simplified JSON-RPC 2.0 (without jsonrpc field)

import Foundation

// MARK: - NativeRPCHost

/// The central RPC host that manages services, handles message routing,
/// and maintains connections to clients.
///
/// Protocol: Simplified JSON-RPC 2.0
/// - Request:      {"id": "1", "method": "service.method", "params": {...}}
/// - Response:     {"id": "1", "result": ...}
/// - Error:        {"id": "1", "error": {"code": -32601, "message": "..."}}
/// - Notification: {"method": "service.event", "params": {...}}
///
/// Usage:
/// ```swift
/// let host = NativeRPCHost()
///
/// // Register services
/// host.register(AppService())
/// host.register(UserService())
///
/// // Connect to Flutter
/// let connection = FlutterMethodChannelConnection(channel: channel)
/// host.addConnection(connection)
/// ```
///
/// Note: Marked `@unchecked Sendable` because all mutable state is protected
/// by the internal `queue` (concurrent DispatchQueue with barrier writes).
/// Safety invariant: All mutations use `.barrier` flag, reads use `queue.sync`.
public final class NativeRPCHost: NativeRPCHostProtocol, @unchecked Sendable {
    
    // MARK: - Properties
    
    /// Registered services by name
    private var services: [String: NativeRPCService] = [:]
    
    /// Active connections
    private var connections: [NativeRPCConnection] = []
    
    /// Event subscriptions: [eventFullName: Set<connectionId>]
    /// where eventFullName = "service.event"
    private var subscriptions: [String: Set<String>] = [:]
    
    /// Serial queue for thread-safe operations
    private let queue = DispatchQueue(label: "com.nativerpc.host", attributes: .concurrent)
    
    /// JSON encoder for messages
    private let encoder = JSONEncoder()
    
    // MARK: - Initialization
    
    public init() {
        encoder.outputFormatting = .sortedKeys
    }
    
    // MARK: - Service Registration
    
    /// Register a service with the host.
    ///
    /// - Parameter service: The service to register
    public func register(_ service: NativeRPCService) {
        let name = service.name
        
        queue.async(flags: .barrier) { [weak self] in
            guard let self = self else { return }
            
            if self.services[name] != nil {
                print("[NativeRPC] Warning: Service '\(name)' is already registered, replacing...")
            }
            
            self.services[name] = service
            service.onRegistered(host: self)
            
            print("[NativeRPC] Registered service: \(name)")
        }
    }
    
    /// Register multiple services at once
    public func register(_ services: NativeRPCService...) {
        for service in services {
            register(service)
        }
    }
    
    /// Unregister a service by name
    public func unregister(name: String) {
        queue.async(flags: .barrier) { [weak self] in
            guard let self = self else { return }
            
            if let service = self.services.removeValue(forKey: name) {
                service.destroy()
                // Remove all subscriptions for this service
                self.subscriptions = self.subscriptions.filter { !$0.key.hasPrefix("\(name).") }
                print("[NativeRPC] Unregistered service: \(name)")
            }
        }
    }
    
    /// Get a registered service by name
    public func getService(_ name: String) -> NativeRPCServiceProtocol? {
        var result: NativeRPCServiceProtocol?
        queue.sync {
            result = services[name]
        }
        return result
    }
    
    // MARK: - Connection Management
    
    /// Add a connection to the host
    public func addConnection(_ connection: NativeRPCConnection) {
        queue.async(flags: .barrier) { [weak self] in
            guard let self = self else { return }
            
            self.connections.append(connection)
            connection.onMessage = { [weak self] data in
                self?.handleIncomingMessage(data, from: connection)
            }
            
            print("[NativeRPC] Added connection: \(connection.id)")
        }
    }
    
    /// Remove a connection from the host
    public func removeConnection(_ connection: NativeRPCConnection) {
        queue.async(flags: .barrier) { [weak self] in
            guard let self = self else { return }
            
            self.connections.removeAll { $0.id == connection.id }
            
            // Remove all subscriptions for this connection
            for (event, var connectionIds) in self.subscriptions {
                connectionIds.remove(connection.id)
                self.subscriptions[event] = connectionIds
            }
            
            print("[NativeRPC] Removed connection: \(connection.id)")
        }
    }
    
    // MARK: - Message Handling
    
    /// Handle an incoming message from a connection
    private func handleIncomingMessage(_ data: Data, from connection: NativeRPCConnection) {
        Task {
            do {
                let message = try NativeRPCMessageParser.parse(data)
                
                switch message {
                case .call(let request):
                    await handleCallRequest(request, from: connection)
                    
                case .subscribe(let request):
                    handleSubscribe(request, from: connection)
                    
                case .unsubscribe(let request):
                    handleUnsubscribe(request, from: connection)
                }
            } catch let error as NativeRPCError {
                sendError(id: "unknown", error: error, to: connection)
            } catch {
                let rpcError = NativeRPCError.parseError(error.localizedDescription)
                sendError(id: "unknown", error: rpcError, to: connection)
            }
        }
    }
    
    /// Handle a call request
    /// Request format: {"id": "1", "method": "service.method", "params": {...}}
    private func handleCallRequest(_ request: NativeRPCRequest, from connection: NativeRPCConnection) async {
        let serviceName = request.service
        let methodName = request.methodName
        
        // Find the service
        guard let service = services[serviceName] else {
            let error = NativeRPCError.serviceNotFound(serviceName)
            sendError(id: request.id, error: error, to: connection)
            return
        }
        
        // Check if method exists
        guard service.canHandle(method: methodName) else {
            let error = NativeRPCError.methodNotFound(methodName, service: serviceName)
            sendError(id: request.id, error: error, to: connection)
            return
        }
        
        // Get params (can be dict or array or nil)
        let params = request.params
        
        do {
            // Call the method
            let result = try await service.handleCall(method: methodName, params: params)
            
            // Send success response: {"id": "1", "result": ...}
            let response = NativeRPCResponse(id: request.id, result: result)
            sendResponse(response, to: connection)
        } catch let error as NativeRPCError {
            sendError(id: request.id, error: error, to: connection)
        } catch {
            let rpcError = NativeRPCError.internalError(error.localizedDescription)
            sendError(id: request.id, error: rpcError, to: connection)
        }
    }
    
    /// Handle a subscribe request
    /// Request format: {"id": "1", "method": "rpc.subscribe", "params": {"event": "service.event"}}
    private func handleSubscribe(_ request: NativeRPCSubscribeRequest, from connection: NativeRPCConnection) {
        let serviceName = request.service
        let eventName = request.eventName
        let eventFullName = request.event  // "service.event"
        
        // Validate service exists
        guard let service = services[serviceName] else {
            let error = NativeRPCError.serviceNotFound(serviceName)
            sendError(id: request.id, error: error, to: connection)
            return
        }
        
        // Validate event is declared
        guard service.definitionContainer.hasEvent(eventName) else {
            let error = NativeRPCError.eventNotDeclared(eventName, service: serviceName)
            sendError(id: request.id, error: error, to: connection)
            return
        }
        
        queue.async(flags: .barrier) { [weak self] in
            guard let self = self else { return }
            
            // Add subscription
            var subscribers = self.subscriptions[eventFullName] ?? []
            let isFirstSubscriber = subscribers.isEmpty
            subscribers.insert(connection.id)
            self.subscriptions[eventFullName] = subscribers
            
            // Notify service if this is the first subscriber
            if isFirstSubscriber {
                service.onStartObserving(event: eventName)
            }
            
            // Send success response: {"id": "1", "result": true}
            let response = NativeRPCResponse(id: request.id, result: true)
            self.sendResponse(response, to: connection)
        }
    }
    
    /// Handle an unsubscribe request
    /// Request format: {"id": "1", "method": "rpc.unsubscribe", "params": {"event": "service.event"}}
    private func handleUnsubscribe(_ request: NativeRPCUnsubscribeRequest, from connection: NativeRPCConnection) {
        let serviceName = request.service
        let eventName = request.eventName
        let eventFullName = request.event  // "service.event"
        
        queue.async(flags: .barrier) { [weak self] in
            guard let self = self else { return }
            
            var subscribers = self.subscriptions[eventFullName] ?? []
            subscribers.remove(connection.id)
            
            let noMoreSubscribers = subscribers.isEmpty
            self.subscriptions[eventFullName] = subscribers
            
            // Notify service if no more subscribers
            if noMoreSubscribers, let service = self.services[serviceName] {
                service.onStopObserving(event: eventName)
            }
            
            // Send success response: {"id": "1", "result": true}
            let response = NativeRPCResponse(id: request.id, result: true)
            self.sendResponse(response, to: connection)
        }
    }
    
    // MARK: - Event Sending (NativeRPCHostProtocol)
    
    /// Send an event notification to all subscribers
    /// Notification format: {"method": "service.event", "params": {...}}
    public func sendEvent(_ notification: NativeRPCNotification) {
        queue.async { [weak self] in
            guard let self = self else { return }
            
            let eventFullName = notification.method  // "service.event"
            
            // Find subscribers for this event
            guard let subscriberIds = self.subscriptions[eventFullName], !subscriberIds.isEmpty else {
                return
            }
            
            // Encode the notification
            guard let data = try? self.encoder.encode(notification) else {
                print("[NativeRPC] Failed to encode notification: \(eventFullName)")
                return
            }
            
            // Send to all subscribed connections
            for connection in self.connections {
                if subscriberIds.contains(connection.id) {
                    connection.send(data)
                }
            }
        }
    }
    
    /// Convenience method to send event with service, event name, and params
    public func sendEvent(service: String, event: String, params: Any? = nil) {
        let notification = NativeRPCNotification(service: service, event: event, params: params)
        sendEvent(notification)
    }
    
    // MARK: - Response Helpers
    
    private func sendResponse(_ response: NativeRPCResponse, to connection: NativeRPCConnection) {
        guard let data = try? encoder.encode(response) else {
            print("[NativeRPC] Failed to encode response")
            return
        }
        connection.send(data)
    }
    
    private func sendError(id: String, error: NativeRPCError, to connection: NativeRPCConnection) {
        let response = NativeRPCErrorResponse(id: id, error: error)
        guard let data = try? encoder.encode(response) else {
            print("[NativeRPC] Failed to encode error response")
            return
        }
        connection.send(data)
    }
    
    // MARK: - Lifecycle
    
    /// Notify all services that app entered foreground
    public func onAppForeground() {
        queue.sync {
            for service in services.values {
                service.onForeground()
            }
        }
    }
    
    /// Notify all services that app entered background
    public func onAppBackground() {
        queue.sync {
            for service in services.values {
                service.onBackground()
            }
        }
    }
    
    /// Shutdown the host and all services
    public func shutdown() {
        queue.async(flags: .barrier) { [weak self] in
            guard let self = self else { return }
            
            // Destroy all services
            for service in self.services.values {
                service.destroy()
            }
            self.services.removeAll()
            
            // Close all connections
            for connection in self.connections {
                connection.close()
            }
            self.connections.removeAll()
            
            // Clear subscriptions
            self.subscriptions.removeAll()
            
            print("[NativeRPC] Host shutdown complete")
        }
    }
    
    // MARK: - Introspection
    
    /// Get list of all registered service names
    public func getServiceNames() -> [String] {
        var result: [String] = []
        queue.sync {
            result = Array(services.keys)
        }
        return result
    }
    
    /// Get service info for introspection
    public func getServiceInfo(_ name: String) -> ServiceInfo? {
        var result: ServiceInfo?
        queue.sync {
            if let service = services[name] {
                result = ServiceInfo(
                    name: name,
                    methods: service.definitionContainer.getMethodNames(),
                    events: Array(service.definitionContainer.getEventNames()),
                    constants: service.getConstants().keys.map { $0 }
                )
            }
        }
        return result
    }
}

// MARK: - Supporting Types

/// Service information for introspection
public struct ServiceInfo {
    public let name: String
    public let methods: [String]
    public let events: [String]
    public let constants: [String]
}
